In [71]:
# Import Libraries #
import pandas as pd
import numpy as np
import os

In [72]:
###Python shortcut to folder path###
path = r'C:\Users\hazem\01-2023 Instacart Basket Analysis - Master Folder'

In [73]:
## Importing Data Set ##
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [74]:
## Importing Data Set ##
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Find missing values

In [75]:
## Finding missing values ##
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [76]:
## Show missing values for specific column ## 
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [77]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Addressing Missing Values

In [78]:
df_prods.shape

(49693, 5)

In [79]:
## Dropping missing values (apparently) ## 
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [80]:
df_prods_clean.shape

(49677, 5)

# Finding duplicate values

In [81]:
## Finding duplicate values in data frame ##
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [82]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [83]:
df_prods_clean.shape

(49677, 5)

In [84]:
## Dropping duplicate values ##
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [85]:
df_prods_clean_no_dups.shape

(49672, 5)

In [86]:
## Exporting Final data frame to csv ##
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5 Data Consistency Checks

In [87]:
df_ords.columns

Index(['Unnamed: 0', 'order_id', 'user_id', 'order_number',
       'orders_day_of_week', 'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [88]:
df_ords.drop(columns = ['Unnamed: 0'])

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [89]:
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [90]:
## Step 2 ##
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [91]:
#Nothing unsual with df_ords statistics ##

In [92]:
## Step 3 ##
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
    
    ## No mixed type data found ##

In [93]:
## Step 5 ##
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [94]:
df_ords.head(30)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   1187899        1            11                   4                  8   
11   2168274        2             1                   2                 11   
12   1501582        2             2                   5                 10   
13   1901567        2             3                   1                 10   
14    738281        2             4                   2                 10   
15   1673511        2             5                   3                 11   
16   1199898        2             6                   2                  9   
17   3194192        2             7                   2                 12   
18    788338        2             8                   1                 15   
19   1718559        2             9                   2                  9   
20   1447487        2            10                   1                 11   
21   1402090        2            11                   1                 10   
22   3186735        2            12                   1                  9   
23   3268552        2            13                   4                 11   
24    839880        2            14                   3                 10   
25   1492625        2            15                   1                 11   
26   1374495        3             1                   1                 14   
27    444309        3             2                   3                 19   
28   3002854        3             3                   3                 16   
29   2037211        3             4                   2                 18   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0  
11                     NaN  
12                    10.0  
13                     3.0  
14                     8.0  
15                     8.0  
16                    13.0  
17                    14.0  
18                    27.0  
19                     8.0  
20                     6.0  
21                    30.0  
22                    28.0  
23                    30.0  
24                    13.0  
25                    30.0  
26                     NaN  
27                     9.0  
28                    21.0  
29                    20.0

##All the missing values in column days_since_prior_order with total 206209 missing values, after checking, each user_id has
one transaction with missing value (the 1st transaction for each user_id in df ##

In [95]:
df_ords.shape

(3421083, 6)

In [96]:
## Step 6 ##
df_ords['days_since_prior_order'].fillna(1.111484e+01, inplace=True)

In [99]:
df_ords.isnull().sum()

order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

##The missing values are almost 200,000 rows, that's alot of data compared to the data frame, we cannot just dropped them out, so i chose to replace it with the mean value, i found it more reasonable##

In [100]:
## Step 7 ##
df_dups = df_ords[df_ords.duplicated()]

##No duplicate found##

In [102]:
## Exporting Final data frame to csv ##
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))